### ISSUE TITLE
NullPointerException warning in log from WebInfConfiguration after upgrade to 9.4.20.v20190813

### ISSUE THREAD

**Issue by jhannes** [eclipse/jetty.project#4173]
After upgrading to 9.4.20.v20190813 or 9.4.21.v20190926, I get the following WARN in the log:

```
2019-10-08 23:47:23.578:WARN:oejw.WebInfConfiguration:main: Can't generate resourceBase as part of webapp tmp dir name: java.lang.NullPointerException
```

The cause of this warning is when using `webAppContext.setBaseResource(Resource.newClassPathResource("/webapp"))` and packaging the sources into a jar-file. `resource.toURI().getPath()` returns null for jar:-urls (at least on Windows): https://github.com/eclipse/jetty.project/blob/jetty-9.4.21.v20190926/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java#L782

This line was previously `resource.toURI().getPath()` which doesn't return null. https://github.com/eclipse/jetty.project/blob/jetty-9.4.19.v20190610/jetty-webapp/src/main/java/org/eclipse/jetty/webapp/WebInfConfiguration.java#L797

This was probably done to remove a deprecation warning on `Resource.toURL`.

Suggested fix: Replace with `Resource.toURI().toURL().getPath()`, which doesn't return null and doesn't give deprecation warning.

Demonstration repository: https://github.com/jhannes/jetty-9420-nullpointer

**Comment by joakime** [comment]
Unfortunately, `Resource.toURI().toURL().getPath()` (or more accurately `URI.toURL()`) is known to mangle existing URL encodings.

**Comment by janbartel** [comment]
@joakime do you have an alternate suggestion? Otherwise the plugin seems to be unusable on windows.

**Comment by joakime** [comment]
I'll be attempting to replicate on my Windows machine tomm.

**Comment by janbartel** [comment]
@joakime I'm assigning this one to you then, as I don't have windows.

**Comment by joakime** [comment]
Your project does not run on my Windows 10 Pro machine.
It doesn't matter what version of Jetty I use (I went back to Jetty 9.4.7.v20170914), nor what version of Java I used (I used OpenJDK 8u202, Oracle JDK 8u212, and AdoptOpenJDK 11.0.4).

It fails for the same reason on every combination.
The `Resource.newClassPathResource("/webapp")` always returns null.

It's pretty obvious in hind-sight.
You can't have a resource lookup in a classpath of a directory. (this is a limitation of Classpaths in java)
Classpaths lookups are for files, not directories.
You'll need to do is obtain a reference to a file, then resolve the directory that file belongs to.
That will satisfy the Classpath requirements first, then the URI requirements of Resource later.

We do this in all of our own examples as well.

See: 
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/WebAppContextFromClasspath.java
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/ResourceHandlerFromClasspath.java
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/DefaultServletFileServer.java
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/DefaultServletMultipleBases.java
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/XmlEnhancedServer.java
* https://github.com/jetty-project/embedded-jetty-cookbook/blob/master/src/main/java/org/eclipse/jetty/cookbook/MultipartMimeUploadExample.java

Example:

``` java
        // Figure out what path to serve content from
        ClassLoader cl = ResourceHandlerFromClasspath.class.getClassLoader();
        // We look for a file, as ClassLoader.getResource() is not
        // designed to look for directories (we resolve the directory later)
        URL f = cl.getResource("static-root/hello.html");
        if (f == null)
        {
            throw new RuntimeException("Unable to find resource directory");
        }

        // Resolve file to directory
        URI webRootUri = f.toURI().resolve("./").normalize();
        System.err.println("WebRoot is " + webRootUri);

        ResourceHandler handler = new ResourceHandler();
        handler.setBaseResource(Resource.newResource(webRootUri));
```

**Comment by jhannes** [comment]
Sorry about the late response, my github feed has been overwhelmed.

The problem is present on at least Windows and Linux. I've updated the repository with a Dockerfile running on the openjdk:11 Docker base image, which manifests the problem. This also indicates that using a directory as a resource is valid, at least under Java 11.

To reiterate: The server seems starts correctly, but it gives a disconcerting WARN about a NullPointerException, see for example:

https://github.com/jhannes/jetty-9420-nullpointer/blob/9cbdb89eca6777018e0bf67c832dc8919de6b3e5/run-docker.log#L33

**Comment by joakime** [comment]
The server starts, the webapp too, but it has no BaseResource directory.
You have no static content, or accessible resources from the ServletContext while it's running this way.

**Comment by jhannes** [comment]
I don't understand, @joakime. Does the Dockerfile produce a different result for you? I have tried this on Windows, on Docker (Window) and on Linux. In all instances it serves files just fine, but gives this NPE warning.

**Comment by mrsladek** [comment]
Hi, Same regresion on 9.4.21.v20190926.

**Comment by gregw** [comment]
Seams to be a problem we are not really understanding...

**Comment by gregw** [comment]
Looking at the test main, I see nothing wrong with that and I would expect it to work.   We should be able to run an embedded jetty using a resource from a jar as the base resource.

**Comment by janbartel** [comment]
OK I've looked into this. So this NPE does not affect deployment of the webapp. The only thing that it affects is the generation of the name of tmp directory.
@jhannes you can set the name of the tmp dir to avoid the warning in the log.
Meanwhile, I'll make a code change to first try the URI.getPath() way, then fallback to URI.toURL().getPath() way.

**Comment by joakime** [comment]
Opened PR #4268

**Comment by janbartel** [comment]
Fix committed via PR #4268
